### PR TITLE
New low-level proof format

### PR DESCRIPTION
**Pull request by jhoenicke**
This is a large pull request introducing the new low-level proof format.

The patch has become quite big, mostly because it also introduces a new LambdaTerm in Library-SMTLIB.  This required to change all TermWalker to accept this new term. This new lambda term is used for the choose function in the proof format.  We may in the future also remove QuantifiedFormula and instead use an application term on a lambda term, similar to what is planned for SMTLIB-3.0.

Another incompatible change is that theory.term now returns Term instead of ApplicationTerm.  The reason is the automatic normalization of integers/rationals:  The term `(- 15)` is now always represented by a ConstantTerm of Rational instead of application term, even if it was created by the function `Theory.term(FunctionSymbol, Term...)`.  This fixes a problem with strange proofs when the input formula contains `(let ((x 15)) (f (- x)))`. 

The old proofs also changed somewhat. In particular the `@clause` rule has changed and the proved clause has moved to the annotation (and no longer includes the `or`). Also the `@split` rule has been removed and instead `@tautology` and `@res` need to be combined.  Also `and` and `=>` is no longer rewritten to `or` by the TermCompiler and double negation is now implicitly handled by the proof checker.  But most of the other changes are below the clause-level and should therefore not affect the interpolator.  The `:quoted` annotations have been removed.  In the future I may remove more `:quotedCC`, `:quotedLA`, and `:quotedQuant` annotations to simplify the low-level proofs. 

The solver now supports three different proof levels: CLAUSE, FULL and LOWLEVEL.  There is a new option `:proof-level` to choose the proof level.  The interpolator needs CLAUSE or FULL proofs, as the low-level proofs do not include information about the partition and they introduce a lot more "very mixed" terms that are bad for interpolation.  Proofs are generated in the old format (either CLAUSE or FULL).  Low-level proofs, are then generated from FULL proofs, by the `ProofSimplifier`.

@danieldietsch  I included you, so that you are aware of the changes, especially the incompatibilities that may apply to Ultimate:  There may be some compilation issues:  Term manipulating classes may need to be adapted for LambdaTerm (especially TermWalker), some casts to ApplicationTerm may need to be manually introduced.  Also SMTLIBOptions has been renamed to SMTLIBConstants.
It would also be great if you can test if this breaks other things in Ultimate.

Known Bug: FormulaLet on formulas in new proof format can lead to problems with define-fun scope in the new proof format (that FormulaLet doesn't know about).

### CODE REVIEW

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/convert/Clausifier.java [hunk 1]
```diff
@@ -259,15 +261,40 @@ public void perform() {
...
+						final Term orIntro = mTracker.tautology(t.term("or", term, t.term("not", p)),
+								ProofConstants.AUX_OR_POS);
+						final Term split = mTracker.resolution(mAxiom, orIntro);
+						pushOperation(new AddAsAxiom(split, mSource));
+					}
+					return;
+				} else if (positive && at.getFunction() == t.mAnd) {
+					// the axioms added below already imply the auxaxiom clauses.
+					setTermFlags(term, oldFlags | assertedFlag | auxFlag);
+					// A negated or is an and of negated formulas. Hence assert all negated
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/convert/Clausifier.java:279]: Remove comment.

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/MinimalProofChecker.java [hunk 2]
```diff
@@ -0,0 +1,1076 @@
...
+	 */
+	ProofRules mProofRules;
+
+	/**
+	 * The proof cache. It maps each converted proof to the clause it proves.
+	 */
+	HashMap<Term, ProofLiteral[]> mCacheConv;
+
+	/**
+	 * The proof cache. It maps each converted proof to the clause it proves.
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/MinimalProofChecker.java:101]: Fix comment

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/ProofRules.java [hunk 3]
```diff
@@ -0,0 +1,1133 @@
...
+	public final static String EXISTSE = "exists-";
+	// equality chains of length >=3
+	public final static String EQI = "=+";
+	public final static String EQE = "=-";
+	public final static String DISTINCTI = "distinct+";
+	public final static String DISTINCTE = "distinct-";
+
+	public final static String ITE1 = "ite1";
+	public final static String ITE2 = "ite2";
+	public final static String REFL = "refl";
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/ProofRules.java:76]: The text documents suggest that refl is not part of the new proof rules and should be replaced with trans.

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/MinimalProofChecker.java [hunk 4]
```diff
@@ -0,0 +1,1076 @@
...
+			return new ProofLiteral[] { new ProofLiteral(theory.term(SMTLIBConstants.EQUALS, modDef, arg), true),
+					new ProofLiteral(theory.term(SMTLIBConstants.EQUALS, divisor, zero), true) };
+		}
+		case ":" + ProofRules.SELECTSTORE1: {
+			if (!theory.getLogic().isArray()) {
+				throw new AssertionError();
+			}
+			assert annots.length == 1;
+			final Term[] params = (Term[]) annots[0].getValue();
+
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/MinimalProofChecker.java:928]: Is there a reason why you stop checking params.length?

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/ProofSimplifier.java [hunk 5]
```diff
@@ -0,0 +1,4175 @@
...
+			return proofs.get(0);
+		}
+	}
+
+	/**
+	 * Check the tautology that introduces a forall.
+	 *
+	 * @param clause     the clause to check
+	 * @param skolemFuns the Skolemization used in the tautology.
+	 * @return true iff the clause is well-formed.
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/ProofSimplifier.java:223]: Adapt

**Review by tanjaschindler — COMMENTED** [review]
Looks fine to me. The new proof checker is nice (and much simpler than the old one.)
I didn't read the ProofSimplifier in detail, though.
I added some minor comments. I'd recommend to make sure that the text documents and the code match.

**Review by henkele — APPROVED** [review]

**Pull request merged**